File: leetcode/c++/Binary_Search_Tree_Iterator.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class BSTIterator {
private:
    stack<TreeNode *> st;
    TreeNode *p = nullptr;
public:
    BSTIterator(TreeNode *root) {
        p = root;
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        while(p || !st.empty()){
            if(p){
                st.push(p);
                p = p->left;
            }else{
                return true;
            }
        }
        
        return false;
    }

    /** @return the next smallest number */
    int next() {
        int val;

        p = st.top();
        st.pop();
        val = p->val;
        p = p->right;
        
        return val;
    }
};

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = BSTIterator(root);
 * while (i.hasNext()) cout << i.next();
 */

File: leetcode/java/BinarySearchTreeIterator.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class BSTIterator {
    private final Deque<TreeNode> stack = new ArrayDeque<>();
    private TreeNode p;

    public BSTIterator(TreeNode root) {
        p = root;
        while(p != null){
            stack.push(p);
            p = p.left;
        }
    }
    
    

    /** @return whether we have a next smallest number */
    public boolean hasNext() {

        return !stack.isEmpty();
    }

    /** @return the next smallest number */
    public int next() {
        TreeNode t = stack.pop();
        if(t.right != null) {
            p = t.right;
            while(p != null){
                stack.push(p);
                p = p.left;
            }
        }
        return t.val;
    }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */
